[RF-DOCS] Active Record Callbacks [ci-skip] (#5)
Ridhwana/rails: guides/source/active_record_callbacks.md

Active Record Callbacks
=======================

This guide teaches you how to hook into the life cycle of your Active Record
objects.

After reading this guide, you will know:

> * When certain events occur during the life of an Active Record object

Reviewer: ```suggestion
* When certain events occur during the life of an Active Record object.
```

* How to create callback methods that respond to events in the object life
  cycle.
* How to create special classes that encapsulate common behavior for your
  callbacks.

--------------------------------------------------------------------------------

The Object Life Cycle
---------------------

During the normal operation of a Rails application, objects may be created,
updated, and destroyed. Active Record provides hooks into this *object life
cycle* so that you can control your application and its data.

> Callbacks allow you to trigger logic before or after an alteration of an

Reviewer 2: These sentences feel almost like saying the same thing, so they can probably be condensed into one sentence. Such as: Callbacks are methods that get called at certain moments of an
object's life cycle allowing you to trigger logic before or after an alteration of an
object's state.

Congratulations!

> ```

Reviewer: I can't comment on the actual line, but I think there's no need to have the instance variable `@baby` in there, for the sake of the callback example.

Author: ✅

As you will see, there are many life cycle events and you can choose to hook
into any of these either before, after, or even around them.

Callbacks Overview
Callback Registration
------------------

Callbacks are methods that get called at certain moments of an object's life cycle. With callbacks it is possible to write code that will run whenever an Active Record object is created, saved, updated, deleted, validated, or loaded from the database.
To use the available callbacks, you need to implement and register them.
Implementation can be done in a multitude of ways like using ordinary methods,
blocks and procs, or defining custom callback objects using callback classes.
Let's go through each of these implementation techniques.

### Callback Registration

In order to use the available callbacks, you need to register them. You can implement the callbacks as ordinary methods and use a macro-style class method to register them as callbacks:
You can implement the callbacks as a **macro-style method that calls an ordinary
method** for registration.

```ruby
class User < ApplicationRecord

>   validates :login, :email, presence: true

Author: In the Callback Registration Section we discuss the different ways to register and implement a callback, hence I've updated the example to be the same across all the different ways so that they're comparable and easier to follow. 

In addition, a `login` attribute/column didn't make complete sense to me so I used something more common like a `username` column.

        self.username = email
      end
    end
end
```

The **macro-style class methods can also receive a block**. Consider using this
style if the code inside your block is so short that it fits in a single line:

```ruby
class User < ApplicationRecord
  validates :username, :email, presence: true

  before_validation do
    self.username = email if username.blank?
  end
end
```

Alternatively, you can **pass a proc to the callback** to be triggered.

```ruby
class User < ApplicationRecord
  validates :username, :email, presence: true

  before_validation ->(user) { user.username = user.email if user.username.blank? }
end
```

> Lastly, you can define **a custom callback object**, which we will cover later
> in more detail [below](#callback-classes).

Reviewer: I think "later in more detail" and "below" are sort of duplicates of the same "idea", that we'll talk about it more down the guide, but it also felt slightly odd to say that and follow-up with an example, so here's a small suggestion:

```suggestion
Lastly, you can define **a custom callback object**, as shown below. We will
cover these [later in more detail](#callback-classes).
```

We should also consider renaming that section to `Callback Objects` instead of `Callback Classes`, to keep it consistent across the board. (the API also seems to use "objects", because in reality it can be anything that responds to those methods, not necessarily a "class")

```ruby
class User < ApplicationRecord
  validates :username, :email, presence: true

  before_validation AddUsername
end

class AddUsername
  def self.before_validation(record)
    if record.username.blank?
      record.username = record.email
    end
  end
end
```

> Callbacks can also be registered to only fire on certain life cycle events, this

Reviewer: I wonder if this warrants a sub-section / title to make it stand out a little more, since `:on` is commonly used but not documented anywhere else in the guide?

I also thought about potentially moving it lower in the guide, but I'm not so sure about that, or where exactly,  (perhaps after we've talked about the available callbacks, or near the conditional callbacks with if/unless) so maybe just considering a sub-section will be enough.

Author: Makes sense, added a subheading ✅

[`after_commit`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html#method-i-after_commit
[`after_rollback`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html#method-i-after_rollback
[`after_save`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-after_save
[`after_validation`]:
    https://api.rubyonrails.org/classes/ActiveModel/Validations/Callbacks/ClassMethods.html#method-i-after_validation
[`around_create`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-around_create
[`around_save`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-around_save
[`before_create`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-before_create
[`before_save`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-before_save
[`before_validation`]:
    https://api.rubyonrails.org/classes/ActiveModel/Validations/Callbacks/ClassMethods.html#method-i-before_validation

> There are examples below that show how to use these callbacks. We've grouped

Author: I'm most uncertain about the sections below. I wanted to show more examples for the Available Callbacks Sections, I tried a few different ways to do that but eventually settled on subsections to group and describe the callbacks with examples. I'm open to feedback if anyone thinks there's a better way to do this.

Reviewer: I think they're good and it's nice to have extra examples... the only other thought I had was to maybe show all the available types and then come up with an example(s) section that has a few different variations, even mixing different callbacks if that helps.... but I don't have much of a strong opinion one way or another, I think what you have here is already a great improvement over what's currently out there.

them by the operation they are associated with, however they can be used in any
combination.

#### Validation Callbacks

Validation callbacks are triggered by the `valid?` method. They are called

> before and after the validation phase.

Reviewer: It might be a little tricky to say that they're triggered by a certain method, because in fact they can be triggered in a few different ways/calls: `valid?`, `validate`, `create`, `update`, `save` (that I can remember, plus their respective `!` methods when applicable); these would all trigger validations, among other things.

Maybe we could expand to something like this: (feel free to tweak, just trying to include all that info here.)

```suggestion
Validation callbacks are triggered whenever the record is validated directly via
the `valid?` or `validate` methods, or indirectly via `create`, `update`, or
`save`. They are called before and after the validation phase.
```

We might also want to consider linking to the API of those methods.

_Update:_ I just realized we have a section on "running callbacks" which mentions a list of methods that would trigger them... but not which methods trigger which callbacks... which I think is sort of what we were trying to get to here.

Maybe we just point people there? Or maybe it makes sense to keep a smaller list of the most common methods like I mentioned above here.

Author: I like the idea of showing which methods trigger which callbacks as I don't think it's very obvious. Hence, it feels like it makes sense to keep a smaller list of the most common methods like you did in your initial suggested change above. After seeing the changes, let me know If its looks fine.

```ruby
class User < ApplicationRecord
  validates :name, presence: true
  before_validation :normalize_name
  after_validation :check_errors

  private

  def normalize_name
    self.name = name.downcase.titleize if name.present?
    Rails.logger.info("Name normalized to #{name}")
  end

  def check_errors
    if errors.any?
      Rails.logger.error("Validation failed: #{errors.full_messages.join(', ')}")
    end

>   end

Reviewer: Let's remove the blank line after `private`, and indent:

```suggestion
    def normalize_name
      self.name = name.downcase.titleize if name.present?
      Rails.logger.info("Name normalized to #{name}")
    end

    def check_errors
      if errors.any?
        Rails.logger.error("Validation failed: #{errors.full_messages.join(', ')}")
      end
    end
```

end
```

```irb
irb> user = User.new(name: "", email: "john.doe@me.com", password: "abc123456")

> => #<User id: nil, email: "john.doe@me.com", created_at: nil, updated_at: nil, name: "">

Reviewer: ```suggestion
irb> user = User.new(name: "", email: "john.doe@example.com", password: "abc123456")
=> #<User id: nil, email: "john.doe@example.com", created_at: nil, updated_at: nil, name: "">
```

irb> user.valid?
Name normalized to
Validation failed: Name can't be blank
=> false
```

#### Save Callbacks

Save callbacks are triggered by the `save` method. They are called before, after

> and around the object is saved.

Reviewer 2: Should there be a 'when' in here? : They are called before, after and around **when** the object is saved.

> and around the object is saved.

Reviewer: Similar to the validate callbacks section: these would be triggered by `create`, `update`, and/or `save`.

Also let's use the Oxford comma: `before, after, and around`.

Maybe something like this? (again, just an idea)

```suggestion
Save callbacks are triggered whenever the record is persisted (i.e. "saved") to
the underlying database, via the `create`, `update`, or `save` methods. They are
called before, after and around the object is saved.
```

Author: ✅

```ruby
class User < ApplicationRecord
  before_save :encrypt_password
  around_save :log_saving

>   after_save :send_welcome_email

Reviewer: ```suggestion
  after_save :update_cache
```

  private

  def encrypt_password
    self.password = BCrypt::Password.create(password)
    Rails.logger.info("Password encrypted for user with email: #{email}")
  end

  def log_saving
    Rails.logger.info("Saving user with email: #{email}")
    yield
    Rails.logger.info("User saved with email: #{email}")
  end

  def update_cache
    Rails.cache.write("user_data", attributes)
    Rails.logger.info("Update Cache")

>   end

Reviewer: Let's remove the blank line after `private`, and indent:

```suggestion
    def encrypt_password
      self.password = BCrypt::Password.create(password)
      Rails.logger.info("Password encrypted for user with email: #{email}")
    end

    def log_saving
      Rails.logger.info("Saving user with email: #{email}")
      yield
      Rails.logger.info("User saved with email: #{email}")
    end

    def update_cache
      Rails.cache.write("user_data", attributes)
      Rails.logger.info("Update Cache")
    end
```

end
```

```irb
irb> user = User.create(name: "Jane Doe", email: "jane.doe@gmail.com")

Password encrypted for user with email: jane.doe@gmail.com
Saving user with email: jane.doe@gmail.com
User saved with email: jane.doe@gmail.com
Update Cache

> => #<User id: 1, email: "jane.doe@gmail.com", created_at: "2024-03-20 16:02:43.685500000 +0000", updated_at: "2024-03-20 16:02:43.685500000 +0000", name: "Jane Doe">

Reviewer: Let's stick to `@example.com` on these:

```suggestion
irb> user = User.create(name: "Jane Doe", email: "jane.doe@example.com")

Password encrypted for user with email: jane.doe@example.com
Saving user with email: jane.doe@example.com
User saved with email: jane.doe@example.com
Update Cache
=> #<User id: 1, email: "jane.doe@example.com", created_at: "2024-03-20 16:02:43.685500000 +0000", updated_at: "2024-03-20 16:02:43.685500000 +0000", name: "Jane Doe">
```

```

#### Create Callbacks

> Create callbacks are triggered by the `create` method. They are called before,

Reviewer 2: Also missing a 'when' here, unless we reorder the words: They are called before, around, and after the object is saved.

> Create callbacks are triggered by the `create` method. They are called before,
> after and around the object is created.

Reviewer: Also similar to validate/save callbacks, this can be triggered by `create`, `save`, and also `update` (on a new / non-persisted record)

```suggestion
Create callbacks are triggered whenever the record is persisted (i.e. "saved")
to the underlying database **for the first time**, in other words, when we're
saving a new record, via the `create`, `save`, or `update` methods. They are
called before, after and around the object is created.
```

Author: ✅

```ruby
class User < ApplicationRecord
  before_create :set_default_role
  around_create :log_creation
  after_create :send_welcome_email

  private

  def set_default_role
    self.role = "user"
    Rails.logger.info("User role set to default: user")
  end

  def log_creation
    Rails.logger.info("Creating user with email: #{email}")
    yield
    Rails.logger.info("User created with email: #{email}")
  end

  def send_welcome_email
    UserMailer.welcome_email(self).deliver_later
    Rails.logger.info("User welcome email sent to: #{email}")

>   end

Reviewer: Let's remove the blank line after `private`, and indent:

```suggestion
    def set_default_role
      self.role = "user"
      Rails.logger.info("User role set to default: user")
    end

    def log_creation
      Rails.logger.info("Creating user with email: #{email}")
      yield
      Rails.logger.info("User created with email: #{email}")
    end

    def send_welcome_email
      UserMailer.welcome_email(self).deliver_later
      Rails.logger.info("User welcome email sent to: #{email}")
    end
```

end
```

```irb
irb> user = User.create(name: "John Doe", email: "john.doe@gmail.com")

User role set to default: user
Creating user with email: john.doe@gmail.com
User created with email: john.doe@gmail.com
User welcome email sent to: john.doe@gmail.com

> => #<User id: 10, email: "john.doe@gmail.com", created_at: "2024-03-20 16:19:52.405195000 +0000", updated_at: "2024-03-20 16:19:52.405195000 +0000", name: "John Doe">

Reviewer: ```suggestion
irb> user = User.create(name: "John Doe", email: "john.doe@example.com")

User role set to default: user
Creating user with email: john.doe@example.com
User created with email: john.doe@example.com
User welcome email sent to: john.doe@example.com
=> #<User id: 10, email: "john.doe@example.com", created_at: "2024-03-20 16:19:52.405195000 +0000", updated_at: "2024-03-20 16:19:52.405195000 +0000", name: "John Doe">
```

* [`after_save`][]
* [`after_commit`][] / [`after_rollback`][]

[`after_update`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-after_update
[`around_update`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-around_update
[`before_update`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-before_update

WARNING. The `after_save` callback is triggered on both a create and update.
However, it consistently executes after the more specific callbacks
`after_create` and `after_update`, regardless of the sequence in which the macro
calls were made.

We've already covered
[validation](active_record_callbacks.html#validation-callbacks),
[save](active_record_callbacks.html#save-callbacks) and
[commit](active_record_callbacks.html#commit-callbacks) callbacks.
`after_commit` / `after_rollback` examples can be found

> [here](active_record_callbacks.html#after-commit-and-after-rollback).

Reviewer: Let's simplify the linking with anchors only, since it's on the same doc. We can also remove the `commit` portion:

```suggestion
We've already covered [validation](#validation-callbacks) and
[save](#save-callbacks) callbacks. `after_commit` / `after_rollback` examples
can be found [here](#after-commit-and-after-rollback).
```

#### Update Callbacks

```ruby
class User < ApplicationRecord
  before_update :check_role_change
  around_update :log_updating
  after_update :send_update_email

  private

  def check_role_change
    if role_changed?
      Rails.logger.info("User role changed to #{role}")
    end
  end

  def log_updating
    Rails.logger.info("Updating user with email: #{email}")
    yield
    Rails.logger.info("User updated with email: #{email}")
  end

  def send_update_email
    UserMailer.update_email(self).deliver_later
    Rails.logger.info("Update email sent to: #{email}")

>   end

Reviewer: Let's remove the blank line after `private`, and indent:

```suggestion
    def check_role_change
      if role_changed?
        Rails.logger.info("User role changed to #{role}")
      end
    end

    def log_updating
      Rails.logger.info("Updating user with email: #{email}")
      yield
      Rails.logger.info("User updated with email: #{email}")
    end

    def send_update_email
      UserMailer.update_email(self).deliver_later
      Rails.logger.info("Update email sent to: #{email}")
    end
```

end
```

```irb
irb> user = User.find(1)
=> #<User id: 1, email: "john.doe@gmail.com", created_at: "2024-03-20 16:19:52.405195000 +0000", updated_at: "2024-03-20 16:19:52.405195000 +0000", name: "John Doe", role: "admin" >

irb> user.update(role: "admin")
User role changed to admin
Updating user with email: john.doe@gmail.com
User updated with email: john.doe@gmail.com

> Update email sent to: john.doe@gmail.com

Reviewer: ```suggestion
irb> user = User.find(1)
=> #<User id: 1, email: "john.doe@example.com", created_at: "2024-03-20 16:19:52.405195000 +0000", updated_at: "2024-03-20 16:19:52.405195000 +0000", name: "John Doe", role: "admin" >

irb> user.update(role: "admin")
User role changed to admin
Updating user with email: john.doe@example.com
User updated with email: john.doe@example.com
Update email sent to: john.doe@example.com
```

* [`after_destroy`][]
* [`after_commit`][] / [`after_rollback`][]

[`after_destroy`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-after_destroy
[`around_destroy`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-around_destroy
[`before_destroy`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-before_destroy

NOTE: `before_destroy` callbacks should be placed before `dependent: :destroy`
associations (or use the `prepend: true` option), to ensure they execute before
the records are deleted by `dependent: :destroy`.

WARNING. `after_commit` makes very different guarantees than `after_save`,
`after_update`, and `after_destroy`. For example, if an exception occurs in an
`after_save` the transaction will be rolled back and the data will not be
persisted. However, anything that happens `after_commit` can guarantee the
transaction has already been completed and the data was persisted to the

> database. More on [transactional callbacks](#transaction-callbacks) below.

Reviewer: Wondering if we need to keep this, and here. It seems to be related to destroy, but it's not -- it's across all _commit_ callbacks, so maybe it's something that only makes sense under that section specifically?

Author: Yeah, it does seem unrelated. I moved this to the `Transactional Callbacks` ### `after_commit` and `after_rollback` section.

```ruby
class User < ApplicationRecord
  before_destroy :check_admin_count
  around_destroy :log_destroy_operation
  after_destroy :notify_users

  private

  def check_admin_count
    if User.where(admin: true).count == 1 && admin?

>       raise StandardError.new("Cannot delete the last admin user")

Reviewer: Could this be an opportunity to `throw :abort`?

    end
    Rails.logger.info("Checked the admin count")
  end

  def log_destroy_operation
    Rails.logger.info("About to destroy user with ID #{id}")
    yield
    Rails.logger.info("User with ID #{id} destroyed successfully")
  end

>   def notify_users

Reviewer: Note: not sure if we need to change this, but the other example have an actual "mailer" call happening here... perhaps not worth adding, if so just for consistency's sake.

Author: Good catch! It makes sense to keep them consistent, thanks

    Rails.logger.info("Notification sent to other users about user deletion")

>   end

Reviewer: Let's remove the blank line after `private`, and indent:

```suggestion
    def check_admin_count
      if User.where(admin: true).count == 1 && admin?
        raise StandardError.new("Cannot delete the last admin user")
      end
      Rails.logger.info("Checked the admin count")
    end

    def log_destroy_operation
      Rails.logger.info("About to destroy user with ID #{id}")
      yield
      Rails.logger.info("User with ID #{id} destroyed successfully")
    end

    def notify_users
      Rails.logger.info("Notification sent to other users about user deletion")
    end
```

end
```

```irb
irb> user = User.find(1)

> => #<User id: 1, email: "john.doe@gmail.com", created_at: "2024-03-20 16:19:52.405195000 +0000", updated_at: "2024-03-20 16:19:52.405195000 +0000", name: "John Doe", role: "admin" >

Reviewer: ```suggestion
=> #<User id: 1, email: "john.doe@example.com", created_at: "2024-03-20 16:19:52.405195000 +0000", updated_at: "2024-03-20 16:19:52.405195000 +0000", name: "John Doe", role: "admin">
```

irb> user.destroy
Checked the admin count
About to destroy user with ID 1
User with ID 1 destroyed successfully
Notification sent to other users about user deletion
```

`after_commit` / `after_rollback` examples can be found

> [here](active_record_callbacks.html#after-commit-and-after-rollback).

Reviewer: Should we add a similar note to the create callbacks section? It's shown there, we have examples of other callback types, but don't mention these.

Author: ✅

=> #<User id: 1>
```

[`after_find`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-after_find
[`after_initialize`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-after_initialize

### `after_touch`

The [`after_touch`][] callback will be called whenever an Active Record object

> is touched.

Reviewer: I felt like I was wondering what is the meaning of "touching" a bit when I got to this section... not sure if we need to explain it further here, but we could at least point to the `touch` API Documentation for more info?

Author: I've added this - NOTE: You can read more about `touch` [here](https://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-touch)

=> true

> ```

Reviewer: Let's update these examples slightly: use `user` instead of just `u` to keep it consistent with other examples, and use `book` instead of `@book`, no need to use instance vars in an irb/console session.

Author: ✅

* `update!`

> * `valid?`

Reviewer: Let's add `validate` here too!

```suggestion
* `valid?`
* `validate`
```

Additionally, the `after_find` callback is triggered by the following finder
methods:

* `all`
* `first`
* `find`
* `find_by`
* `find_by!`
* `find_by_*`
* `find_by_*!`
* `find_by_sql`
* `last`
* `sole`
* `take`

The `after_initialize` callback is triggered every time a new object of the
class is initialized.

NOTE: The `find_by_*` and `find_by_*!` methods are dynamic finders generated
automatically for every attribute. Learn more about them in the [Dynamic finders

> section](active_record_querying.html#dynamic-finders)

Reviewer: Missing period:

```suggestion
section](active_record_querying.html#dynamic-finders).
```

lead to invalid data.

Halting Execution
-----------------

As you start registering new callbacks for your models, they will be queued for
execution. This queue will include all of your model's validations, the
registered callbacks, and the database operation to be executed.

The whole callback chain is wrapped in a transaction. If any callback raises an
exception, the execution chain gets halted and a ROLLBACK is issued. To
intentionally halt a chain use:

```ruby
throw :abort
```

WARNING. Any exception that is not `ActiveRecord::Rollback` or
`ActiveRecord::RecordInvalid` will be re-raised by Rails after the callback
chain is halted. Additionally, it may break code that does not expect methods
like `save` and `update` (which normally try to return `true` or `false`) to
raise an exception.

NOTE: If an `ActiveRecord::RecordNotDestroyed` is raised within `after_destroy`,
`before_destroy` or `around_destroy` callback, it will not be re-raised and the
`destroy` method will return `false`.

Relational Callbacks
--------------------

Callbacks work through model relationships, and can even be defined by them.

> Suppose an example where a user has many articles. A user's articles should be

Reviewer 2: While reading this, one thing that I kept thinking is that while there are code snippets, there isn't much explaining of when callbacks could be used in the copy of each section, so I love this concrete example here. Would love to see more of these above.

Association Callbacks
---------------------

> Association callbacks are similar to normal callbacks, but they are triggered by
> events in the life cycle of a collection. There are four available association
> callbacks:

Reviewer: ```suggestion
Association callbacks are similar to normal callbacks, but they are triggered by
events in the life cycle of the associated collection. There are four available
association callbacks:
```

* `before_add`
* `after_add`
* `before_remove`
* `after_remove`

You define association callbacks by adding options to the association

> declaration. For example:

Reviewer: Should we make the methods within the examples below private, like the other callback examples we've been showing? (since that's the general recommendation)

Author: ✅

end
```

If a `before_add` callback throws `:abort`, the object does not get added to the
collection. Similarly, if a `before_remove` callback throws `:abort`, the object

> does not get removed from the collection:

Reviewer 2: Should this sentence move below the code example since this position implies that the code will show this example, whereas it shows the code reflecting the first sentence.

```ruby
class Comment < ApplicationRecord
  before_save :filter_content,
    if: proc { forum.parental_control? },
    unless: proc { author.trusted? }
end
```

The callback only runs when all the `:if` conditions and none of the `:unless`
conditions are evaluated to `true`.

Callback Classes
----------------

Sometimes the callback methods that you'll write will be useful enough to be
reused by other models. Active Record makes it possible to create classes that
encapsulate the callback methods, so they can be reused.

> Here's an example where we create a class with an `after_destroy` callback to

Reviewer 2: Another good example of concrete uses to educate about callbacks.

end
```

WARNING. Using both `after_create_commit` and `after_update_commit` with the
same method name will only allow the last callback defined to take effect, as
they both internally alias to `after_commit` which overrides previously defined
callbacks with the same method name.

```ruby

> class User < ApplicationRecord

Reviewer: Let's update this example to remove the `@` instance var in favor of just a local var too.

Author: ✅

### `after_save_commit`

There is also [`after_save_commit`][], which is an alias for using the
`after_commit` callback for both create and update together:

```ruby

> class User < ApplicationRecord

Reviewer: Let's update this example to remove the `@` instance var in favor of just a local var too.

Author: ✅

By default, callbacks will run in the order they are defined. However, when
defining multiple transactional `after_` callbacks (`after_commit`,

> `after_rollback`, etc), the order can be reversed from when they are defined.

Reviewer: This got me wondering if we should document this the other way around now, the default going forward is always in the order they're defined, but previous versions might reverse, and the configuration allows that behavior to be changed.

Author: @carlosantoniodasilva apologies I'm not sure what you mean here, can you expand?

Author: This makes sense, I took a stab at it!

end
```

NOTE: This applies to all `after_*_commit` variations too, such as
`after_destroy_commit`.

This order can be set via configuration:

```ruby
config.active_record.run_after_transaction_callbacks_in_order_defined = false
```

When set to `true` (the default from Rails 7.1), callbacks are executed in the
order they are defined. When set to `false`, the order is reversed, just like in

> the example above.

Reviewer: We should find an opportunity here to link to the configuring guide on `run_after_transaction_callbacks_in_order_defined`

Author: ✅
